### PR TITLE
fix(ingest): bigquery-beta - turning sql parsing off in lineage extraction

### PR DESCRIPTION
**Pull request by treff7es**
- Turning SQL parsing in bigquery beta off by default as it slowed down significantly the lineage generation, and it was not precise enough.
- Collecting report metrics for lineage per project and not globally 
- Fixing table pattern in lineage extraction

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py [hunk 1]
```diff
@@ -56,14 +57,22 @@ class BigQueryV2Config(BigQueryConfig):
...
     # The inheritance hierarchy is wonky here, but these options need modifications.
     project_id: Optional[str] = Field(
         default=None,
         description="[deprecated] Use project_id_pattern instead.",
     )
     storage_project_id: None = Field(default=None, exclude=True)
 
+    lineage_use_sql_parser: bool = Field(
+        default=False,
+        description="Experimental. Use sql parser to resolve view/table lineage. If there is a view being referenced then bigquery sends both the view as well as underlying tablein the references. There is no distinction between direct/base objects accessed. So doing sql parsing to ensure we only use direct objects accessed for lineage.",
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py:73]:
```suggestion
        description="Experimental. Use sql parser to resolve view/table lineage. If there is a view being referenced then BigQuery records both the view and the underlying tables as lineage. Enabling SQL parsing allows us to only include the views/tables that were actually accessed by a query.",
```

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/lineage.py [hunk 2]
```diff
@@ -353,34 +357,62 @@ def _parse_exported_bigquery_audit_metadata(
                     f"Unable to parse audit metadata missing {missing_exported_audit} for {audit_metadata}",
                 )
             else:
-                self.report.num_parsed_audit_entires += 1
+                self.report.num_parsed_audit_entries[event.project_id] = (
+                    self.report.num_parsed_audit_entries.get(event.project_id, 0) + 1
+                )
+                self.report.num_total_audit_entries[event.project_id] = (
+                    self.report.num_total_audit_entries.get(event.project_id, 0) + 1
+                )
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/lineage.py:365]: Ideally the TopKDict would be a subclass of defaultdict. That'd let this code be simplified

```suggestion
                self.report.num_total_audit_entries[event.project_id] += 1
```

File: metadata-ingestion/src/datahub/utilities/stats_collections.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    @staticmethod
+    def _trim_dictionary(big_dict: Dict[str, Any]) -> Dict[str, Any]:
+        if big_dict is not None and len(big_dict) > 10:
+            dict_as_tuples = [(k, v) for k, v in big_dict.items()]
+            sorted_tuples = sorted(dict_as_tuples, key=lambda x: x[1], reverse=True)
+            dict_as_tuples = sorted_tuples[:10]
+            trimmed_dict = {k: v for k, v in dict_as_tuples}
+            trimmed_dict[f"... top(10) of total {len(big_dict)} entries"] = ""
+            print(f"Dropping entries {sorted_tuples[11:]}")
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/stats_collections.py:29]: nit: use logger

File: metadata-ingestion/src/datahub/utilities/stats_collections.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
+
+
+class TopKDict(Dict[_KT, _VT]):
+    """A structure that only prints the top K items from the dictionary. Not lossy."""
+
+    def __init__(self, top_k: int = 10) -> None:
+        super().__init__()
+        self.top_k = 10
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/stats_collections.py:13]:
```suggestion
        self.top_k = top_k
```

File: metadata-ingestion/src/datahub/utilities/stats_collections.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+from typing import Any, Dict, TypeVar, Union
+
+T = TypeVar("T")
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/stats_collections.py:3]: it doesn't look like this is used anywhere?

**Review by shirshanka — APPROVED** [review]
LGTM

**Pull request merged**